pkg/querier/queryrange/querysharding.go

```diff
@@ -0,0 +1,307 @@
+}
+
+func (ast *astMapperware) Do(ctx context.Context, r Request) (Response, error) {
+	conf, err := ast.confs.GetConf(r)
+	// cannot shard with this timerange
+	if err != nil {
+		level.Warn(ast.logger).Log("err", err.Error())
+		return ast.next.Do(ctx, r)
+	}
+
+	shardSummer, err := astmapper.NewShardSummer(int(conf.RowShards), astmapper.VectorSquasher)
+	if err != nil {
+		return nil, err
+	}
+
+	subtreeFolder := astmapper.NewSubtreeFolder()
+
+	strQuery := r.GetQuery()
+	mappedQuery, err := mapQuery(
+		astmapper.NewMultiMapper(
+			shardSummer,
+			subtreeFolder,
+		),
+		strQuery,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	strMappedQuery := mappedQuery.String()
+	level.Debug(ast.logger).Log("msg", "mapped query", "original", strQuery, "mapped", strMappedQuery)
+	mappedASTCounter.Inc()
+
+	return ast.next.Do(ctx, r.WithQuery(strMappedQuery))
+
+}
+
+type queryShard struct {
+	confs  ShardingConfigs
+	next   Handler
+	engine *promql.Engine
+}
+
+func (qs *queryShard) Do(ctx context.Context, r Request) (Response, error) {
+	// since there's no available sharding configuration for this time range,
+	// no astmapping has been performed, so skip this middleware.
+	if _, err := qs.confs.GetConf(r); err != nil {
+		return qs.next.Do(ctx, r)
+	}
+
+	queryable := lazyquery.NewLazyQueryable(&DownstreamQueryable{r, qs.next})
+
+	qry, err := qs.engine.NewRangeQuery(
+		queryable,
+		r.GetQuery(),
+		TimeFromMillis(r.GetStart()),
+		TimeFromMillis(r.GetEnd()),
+		time.Duration(r.GetStep())*time.Millisecond,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+	res := qry.Exec(ctx)
+	extracted, err := FromResult(res)
+	if err != nil {
+		return nil, err
+
+	}
+	return &PrometheusResponse{
+		Status: StatusSuccess,
+		Data: PrometheusData{
+			ResultType: string(res.Value.Type()),
+			Result:     extracted,
+		},
+	}, nil
+}
+
+// shardSplitter middleware will only shard appropriate requests that do not extend past the MinShardingLookback interval.
+// This is used to send nonsharded requests to the ingesters in order to not overload them.
+type shardSplitter struct {
+	codec               Codec
+	MinShardingLookback time.Duration    // delimiter for splitting sharded vs non-sharded queries
+	shardingware        Handler          // handler for sharded queries
+	next                Handler          // handler for non-sharded queries
+	now                 func() time.Time // injectable time.Now
+}
+
+func (splitter *shardSplitter) Do(ctx context.Context, r Request) (Response, error) {
+	cutoff := splitter.now().Add(-splitter.MinShardingLookback)
+	sharded, nonsharded := partitionRequest(r, cutoff)
+
+	return splitter.parallel(ctx, sharded, nonsharded)
+
+}
+
+func (splitter *shardSplitter) parallel(ctx context.Context, sharded, nonsharded Request) (Response, error) {
+	if sharded == nil {
+		return splitter.next.Do(ctx, nonsharded)
+	}
+
+	if nonsharded == nil {
+		return splitter.shardingware.Do(ctx, sharded)
+	}
+
+	nonshardCh := make(chan Response, 1)
```

Comment:
None of these channels are ever closed. In the event of a timeout I think they will remain open forever. This will likely cause a memory leak over time.

Reply:
I think it should be fine, the GC should take care of it actually